library/modules/EventManager.cpp

```diff
@@ -527,12 +533,18 @@ static void manageJobCompletedEvent(color_ostream& out) {
     int32_t tick0 = eventLastTick[EventType::JOB_COMPLETED];
     int32_t tick1 = df::global::world->frame_counter;
 
-    multimap<Plugin*,EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
-    unordered_map<int32_t, df::job*> nowJobs;
-    for ( df::job_list_link* link = &df::global::world->jobs.list; link != nullptr; link = link->next ) {
-        if ( link->item == nullptr )
-            continue;
-        nowJobs.emplace(link->item->id, link->item);
+    multimap<Plugin*, EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
+    std::unordered_map<int32_t, JobCompleteData> nowJobs;
+    nowJobs.reserve(prevJobs.size());
+    for (df::job_list_link* link = &df::global::world->jobs.list; link != nullptr; link = link->next) {
+        if (link->item != nullptr) {
+            df::job& job = *link->item;
```

Comment:
This may be a simple as:
```suggestion
for (auto job : df::global::world->jobs.list)
```

Reply:
That should be a C-style linked list - we don't get to use the fancy range-style-for those types of containers.

Reply:
I literally copied that line from the implementation of `suspendmanager.cpp`: https://github.com/DFHack/dfhack/blob/9e5c4b8ac6f7def545b1d920c1a72e0bb0380cd3/plugins/suspendmanager.cpp#L643

Reply:
our currently implementation of `DFLinkedList` does not satisfy the InputIterator concept requirements and thus can't be used with ranges until someone fixes this. See #4318